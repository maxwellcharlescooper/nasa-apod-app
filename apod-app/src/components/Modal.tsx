import React, { useRef } from "react";
import styled from "styled-components";
import { Apod } from "../types/Apod";
import Styled from "../styles/Styles";
import Functions from "../util/Functions";
import { format, parse } from "date-fns";
import Colors from "../styles/Colors";

const Card = styled(Styled.FlexColumn)`
  width: 700px;
  height: 700px;
  background-color: white;
  color: black;
  align-items: center;
  padding: 20px;
  border-radius: 10px;
`;

const Entry = styled(Styled.FlexRow)`
  margin-bottom: 30px;
`;

const Close = styled(Styled.FlexColumn).attrs({
  "aria-label": "close"
})`
  cursor: pointer;
  display: flex;
  flex-direction: column;

  &:hover {
    background-color: ${Colors.GRAY};
  }
`;

type Props = {
  row: Apod | null;
  closeModal: () => void;
};

const Modal = ({ row, closeModal }: Props) => {
  const ref = useRef();
  Functions.useOnClickOutside(ref, closeModal);
  return (
    <Styled.ModalOverlay>
      <Card ref={ref}>
        <img src={row.url} height="300" />
        <Entry>{row.title}</Entry>
        <Entry>
          {format(parse(row.date, "yyyy-MM-dd", new Date()), "LLLL do, yyyy")}
        </Entry>
        <Entry>{row.explanation}</Entry>
        <Close onClick={closeModal}>
          <Styled.CloseIcon />
        </Close>
      </Card>
    </Styled.ModalOverlay>
  );
};

export default Modal;
