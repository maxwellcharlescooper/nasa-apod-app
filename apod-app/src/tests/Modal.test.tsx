import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import Modal from "../components/Modal";
import { testData } from "../tests/testData";

describe("Modal.tsx", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <Modal
          row={testData[0]}
          closeModal={() => {
            return;
          }}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("closes on click", () => {
    const handleClick = jest.fn();
    render(<Modal row={testData[0]} closeModal={handleClick} />);

    fireEvent.click(screen.getByRole("generic", { name: "close" }));

    expect(handleClick).toBeCalledTimes(1);
  });
});
