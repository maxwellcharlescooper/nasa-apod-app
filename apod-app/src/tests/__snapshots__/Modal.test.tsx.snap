// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.tsx renders correctly 1`] = `
<div
  className="sc-bdfCkl sc-hKgJyM svrZn lnySgQ"
>
  <div
    className="sc-gsTChe sc-eCstFB VpFSV ionEbd"
  >
    <img
      height="300"
      src="https://apod.nasa.gov/apod/image/2306/MavenMars2panel.png"
    />
    <div
      className="sc-bdfCkl sc-jSgtCu svrZn fodjlz"
    >
      MAVEN's Ultraviolet Mars
    </div>
    <div
      className="sc-bdfCkl sc-jSgtCu svrZn fodjlz"
    >
      June 27th, 2023
    </div>
    <div
      className="sc-bdfCkl sc-jSgtCu svrZn fodjlz"
    >
      These two global views of Mars were captured at ultraviolet wavelengths, beyond the spectrum visible to human eyes. Recorded by the MAVEN spacecraft's Imaging Ultraviolet Spectrograph instrument in July 2022 (left) and January 2023, three otherwise invisible ultraviolet bands are mapped into red, green, and blue colors. That color scheme presents the Red Planet's surface features in shades of tan and green. Haze and clouds appear white or blue, while high altitude ozone takes on a dramatic purple hue. On the left, Mars' south polar ice cap is in brilliant white at the bottom but shrinking during the southern hemisphere's summer season. On the right, the northern hemisphere's polar region is seen shrouded in clouds and atmospheric ozone. Known to some as the Mars Atmosphere and Volatile EvolutioN spacecraft, MAVEN has been exploring Mars' tenuous upper atmosphere, ionosphere, and its interactions with the Sun and solar wind since 2014.
    </div>
    <div
      aria-label="close"
      className="sc-gsTChe sc-gKsdJj VpFSV cxStmS"
      onClick={[Function]}
    >
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M18.3 5.71c-.39-.39-1.02-.39-1.41 0L12 10.59 7.11 5.7c-.39-.39-1.02-.39-1.41 0-.39.39-.39 1.02 0 1.41L10.59 12 5.7 16.89c-.39.39-.39 1.02 0 1.41.39.39 1.02.39 1.41 0L12 13.41l4.89 4.89c.39.39 1.02.39 1.41 0 .39-.39.39-1.02 0-1.41L13.41 12l4.89-4.89c.38-.38.38-1.02 0-1.4z"
        />
      </svg>
    </div>
  </div>
</div>
`;
